### ISSUE TITLE
Consider adding "Actors" and the ability to rotate keys

### ISSUE THREAD

**Issue by iamcarbon** [egregore-project/docs#3]
Identifying all actors by their public key prevents an actor from rotating or revoking their key. By pointing owners to a versioned actor record, the system could support rotating and revoking keys. 

A signed trusted central ledger of key revocations could also be used to prove whether a record signed at a point in time was valid.

By establishing a set of trusted actors, a key recovery method could be introduced that allows trusted actors to commit a key rotation transaction through a cryptographically verified quorum.

Perhaps one of these actors is a friend. A T2 chip. Or FIDO key.

**Comment by iamcarbon** [comment]
A potential Schema.

```
Actor record { 
   id      : UUID, 
   type : Person | Service
}

Actor::Key_reset_policy record {
    quorum  : Integer, // The number of trustees that must agree to let the actor reset their key
    trustees : [ ] Trustee
} 

// A trustline establishes trust between two parties (or keys) for a specific purpose.
// 1. To assert our identity (that we control our current private key) 
// 2. That the trustee hold an asset on our behalf and will provide it on demand
// NOTE: by trusting a actor, we inherently trust their currently active key

Trustline record { 
   truster     : Actor,  
   trustee     : Person | Key,   
   acting_as : Verifier of Identity | Holder of Asset
   expires?   : Timestamp
}

Record::Signature record { 
   key    : Cryptography::Key, 
   value : [ ] byte 
} 


Cryptography::Key record { 
  uuid                         : UUID, 
  type.                        : Cryptography::Algorithm,
  revoked?                 : Timestamp, 
  revocation_reason : Rotated | Compromised, 
  owner                       : Actor 
}
```

**Comment by danielcrenna** [comment]
@iamcarbon

I like this idea.

One consideration on the revocation is that I don't think I could legitimately assert that a key was revoked prior to an actor committing a transaction. The revocation list itself would have to have quorum, and then you might have a situation where you can prevent a revocation from "taking" by upsetting that state.

**Comment by iamcarbon** [comment]
Yeah. There's some tricky timing & coordination issues in this proposal. It would also necessitate trusting third parties (e.g. https://blog.cloudflare.com/roughtime/) to validate things happened in a certain order. 

For reliability / availability,  the individual systems should be able to update on their current knowledge of what keys are valid -- and retroactively record to when learning of key revocations. It may be reasonable to add a new flag state to records (RecordFlags = Invalidated). This may also be infeasible with an append only log (if the space cannot be reserved and updated in-place).

**Comment by danielcrenna** [comment]
I'm not against the idea of a mutable log, but it complicates several replication scenarios and removes "free" data provenance features.

I think the existing implementation needs to reach a further milestone before we black-boarding data removal.

Worth pointing out that there are no crypto-economics at play here, so nothing prevents running a private instance where exposing sensitive data becomes a network concern.

Are there ready examples of the need to delete data legally outside of GDPR or is that your primary concern?

**Comment by iamcarbon** [comment]
Agreed that this is a non-inniment-concern for private instances and small syndications that inherently trust each other and communicate peer to peer. 

My future brain is leaping ahead where instances can form a network — and share information and “things” with one another — where the possibility of trust bring compromised - either by a malicious actor or a system a CSV / zero day — becomes inevitable. 

By supporting the right to be forgotten / GDPR, CCAP and DMCA, the system would also allow public providers to participate and establish themselves as trust anchors. 

﻿I also believe that  it remains possible to achieve GDPR and CCAP compliance without loosing data provenance -  by keeping the records tombstone (soft delete marker) along with the records hash - while destroying the records data. Loosing the records DEK is also a solution. 

But yes, future problem!!!
